### DISCUSSION TITLE
Curious warning about dropped aesthetics in stat

### DISCUSSION THREAD

**Discussion by yjunechoe** [ggplot2-extenders/ggplot-extension-club#47] [General]
Not saying that I'm doing a reasonable thing here (I rarely am!) but I'm curious about the warning (in both DEV and CRAN) that I get for a code like:

```r
ggplot(diamonds, aes(cut, group = clarity)) +
  geom_bar(
    aes(fill = cut),
    color = "white"
  )
#> Warning: The following aesthetics were dropped during statistical transformation: fill.
#> ℹ This can happen when ggplot fails to infer the correct grouping structure in the data.
#> ℹ Did you forget to specify a `group` aesthetic or to convert a numerical variable into a factor?
```

![image](https://github.com/user-attachments/assets/5858236f-a78c-45ac-9524-b597a49ae954)

Where I might have expected the above to produce a plot equivalent to:

```r
library(dplyr)

diamonds_counted <- diamonds %>% 
    count(cut, clarity)
    
ggplot(diamonds_counted, aes(cut, n, group = clarity)) +
  geom_col(
    aes(fill = cut),
    color = "white"
  )
```

![image](https://github.com/user-attachments/assets/22bd1eed-adc0-4d53-812e-a34586368f30)

The technicality behind the warning aside (though I'm curious about that as well), the message doesn't seem so informative? It hints:

> ℹ Did you forget to specify a `group` aesthetic or to convert a numerical variable into a factor?

Which is a yes (`group` is specified) and yes (`clarity` is a factor).

Is there a clear explanation for this behavior? (and what a better message could be?)

I *think* this is related to an issue I found a while ago - https://rpubs.com/yjunechoe/stat-count-non-constant-aes - but I think I over-complicated the examples there, so I fail to understand what I wrote after having forgotten about it for so long 😅

**Comment by yjunechoe** [comment]
Also if this feels like an actual issue (at least at the scope of warning message informativeness) I'll polish this up and open an issue for that specifically, but I'm kinda stumped about the scope and proper description of the "problem" atm!

**Comment by teunbrand** [comment]
Simplified example:

``` r
library(ggplot2)

df <- data.frame(
  x = c("A", "B", "B"),
  g = c("X", "X", "Y")
)

ggplot(df, aes(x, fill = x, group = g)) +
  geom_bar(colour = "white")
#> Warning: The following aesthetics were dropped during statistical transformation: fill.
#> ℹ This can happen when ggplot fails to infer the correct grouping structure in
#>   the data.
#> ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
#>   variable into a factor?
```

![](https://i.imgur.com/KdzVLzv.png)

The issue is we group by `g` and thus count (and also attempting to apply fill over) over `x = "A"` and `x = "B"` for `g = "X"`. Because we have 2 fill values for `g = "X"`, we get the non-constant aesthetic warning.  Because you declared the groups manually, you don't get the luxury of imposed group-by-x you'd get by default. The fix would be to declare the groups as an interaction between the two variables:

``` r
ggplot(df, aes(x, fill = x, group = interaction(x, g))) +
  geom_bar(colour = "white")
```

![](https://i.imgur.com/bWP4QdH.png)

<sup>Created on 2024-10-09 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
Oh - I see!

The usecase was a (yet another hacky) `after_scale` transformation of to give a `fill` darkening gradient across `group` of bars within each `x`. Much more trivial to mimic this effect with `alpha` but this looks prettier 😃

```r
library(ggplot2)
library(colorspace)

ggplot(diamonds, aes(cut, group = interaction(cut, clarity))) +
  geom_bar(
    aes(fill = stage(cut,
                     after_scale = ave(fill, x, PANEL, FUN = \(.x) {
                       darken(.x, seq(0.1, 0.9, length.out = length(.x)))
                     }))),
    color = "white",
    show.legend = FALSE
  )
```

![image](https://github.com/user-attachments/assets/95e44376-d696-439a-9bc0-6c41c49f5c51)

**Reply by teunbrand** [reply]
Nice use of `ave()` here!

**Reply by yjunechoe** [reply]
I think I learned the `ave()` trick from you ;)
